### PR TITLE
backport: Merge bitcoin#19082,19227,19206,16756,1883419252,19304,19350,18895,18807,18828

### PR DESCRIPTION
**Pull request by vijaydasmp**
Issue being fixed or feature implemented
Bitcoin Backports

What was done?
Misc v21 Backports, check the commit log to see changes.

How Has This Been Tested?
Compilation and tested unit tests locally

Breaking Changes
Checklist:
 I have performed a self-review of my own code

### CODE REVIEW

File: test/functional/mining_basic.py [hunk 1]
```diff
@@ -89,12 +87,6 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
         coinbase_tx.rehash()
 
         # round-trip the encoded bip34 block height commitment
```
Comment (vijaydasmp) [inline test/functional/mining_basic.py]: removing this comment as well

File: doc/release-notes.md [hunk 2]
```diff
@@ -107,6 +107,17 @@ None
 ### Changes in existing RPCs introduced through bitcoin backports:
 - `walletprocesspsbt` and `walletcreatefundedpsbt` now include BIP 32 derivation paths by default for public keys if we know them. This can be disabled by setting `bip32derivs` to `false`.
```
Comment (vijaydasmp) [inline doc/release-notes.md:108]: This change needs a suggestion

**Comment by vijaydasmp** [comment]
Hello @UdjinM6 , requesting review

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/mining_basic.py [hunk 3]
```diff
@@ -112,7 +113,10 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
         assert_raises_rpc_error(-22, "Block does not start with a coinbase", node.submitblock, bad_block.serialize().hex())
 
         self.log.info("getblocktemplate: Test truncated final transaction")
-        assert_raises_rpc_error(-22, "Block decode failed", node.getblocktemplate, {'data': block.serialize()[:-1].hex(), 'mode': 'proposal'})
+        assert_raises_rpc_error(-22, "Block decode failed", node.getblocktemplate, {
+            'data': block.serialize()[:-1].hex(),
+            'mode': 'proposal',
+        })
```
[the remark this answers is not in the document]
Comment (vijaydasmp) [reply test/functional/mining_basic.py:119]: changes done!

File: doc/release-notes.md [hunk 4]
```diff
@@ -107,6 +107,17 @@ None
...
+  mempool tracks locally submitted transactions until their initial broadcast
+  is acknowledged by a peer. This field returns the count of transactions
+  waiting for acknowledgement.
+
+- `getmempoolentry` and `getrawmempool` with `verbose=true` now return an additional
+  `unbroadcast` field. This indicates  whether initial broadcast of the transaction
+  has been acknowledged by a  peer. `getmempoolancestors` and `getmempooldescendants`
+  are also updated.
+
+
```
[the remark this answers is not in the document]
Comment (vijaydasmp) [reply doc/release-notes.md]: created doc/release-notes-18807.md

**Review by UdjinM6 — APPROVED** [review]
utACK

File: test/functional/mempool_compatibility.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test that mempool.dat is both backward and forward compatible between versions
```
Comment (PastaPastaPasta) [inline test/functional/mempool_compatibility.py:5]: 19153 I don't think we can do this...

In its current form the test is skipped when trying to run it locally. I would think we should at least add a comment or something indicating that this test can't actually be ran...

File: test/functional/test_framework/mininode.py [hunk 6]
```diff
@@ -722,6 +722,8 @@ def on_inv(self, message):
                 # save txid
                 self.tx_invs_received[i.hash] += 1
 
+        super().on_inv(message)
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/mininode.py:725]: I don't see why https://github.com/bitcoin/bitcoin/pull/18807/files#diff-2dd85d481900d4ad19d113d2114861b0134bcd283435e95b18d10adf5ad381a0 couldn't be done and make 18807 complete

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

**Comment by vijaydasmp** [comment]
Hello @UdjinM6  @PastaPastaPasta  wallet_resendwallettransactions.py test seems to be failing , requesting suggestion for this file

**Comment by knst** [comment]
> Hello @UdjinM6 @PastaPastaPasta tried https://github.com/bitcoin/bitcoin/pull/18807/files#diff-2dd85d481900d4ad19d113d2114861b0134bcd283435e95b18d10adf5ad381a0 But the wallet_resendwallettransactions.py test seems to be failing, requesting suggestion for this file

@vijaydasmp please, check this patch: https://github.com/knst/dash/commit/f6ac80cea4e09cf4b6db29fcd93b8be3cf0b2a7f
Dash codebase use slightly different implementation of mocktime.

File: test/functional/test_framework/test_framework.py [hunk 7]
```diff
@@ -436,6 +438,25 @@ def add_nodes(self, num_nodes, extra_args=None, *, rpchost=None, binary=None):
 
         Should only be called once after the nodes have been specified in
         set_test_params()."""
+        def get_bin_from_version(version, bin_name, bin_default):
```
Comment (knst) [inline test/functional/test_framework/test_framework.py:441]:
```
            binary = [get_bin_from_version(v, 'bitcoind', self.options.bitcoind) for v in versions]
```
hm, why this function is not used below as in bitcoin?

**Review by knst — CHANGES_REQUESTED** [review]
IMO better to some prior backports:
 - Merge bitcoin#18628 before bitcoin#19304
 - Merge bitcoin#18334 before bitcoin#19252

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/test_runner.py [hunk 8]
```diff
@@ -69,6 +69,7 @@
 
 TEST_FRAMEWORK_MODULES = [
     "muhash",
+    "blocktools",
```
Comment (UdjinM6) [inline test/functional/test_runner.py:72]: 19082: move this line up (the list should be sorted alphabetically)

File: test/functional/mempool_compatibility.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Currently this test is cant be run
```
Comment (UdjinM6) [inline test/functional/mempool_compatibility.py:5]: 19153: why backporting it then?

File: test/functional/test_framework/messages.py [hunk 10]
```diff
@@ -1780,6 +1782,41 @@ def __repr__(self):
...
+        r += ser_string(self.data)
+        r += struct.pack("<I", self.nHashFuncs)
+        r += struct.pack("<I", self.nTweak)
+        r += struct.pack("<B", self.nFlags)
+        return r
+
+    def __repr__(self):
+        return "msg_filterload(data={}, nHashFuncs={}, nTweak={}, nFlags={})".format(
+            self.data, self.nHashFuncs, self.nTweak, self.nFlags)
+
```
Comment (UdjinM6) [inline test/functional/test_framework/messages.py:1819]: 18334: the merge is messed up, should move this block of new code below `msg_headers2` class definition (20+ lines lower)

File: test/functional/p2p_invalid_messages.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 import struct
 import sys
 
+
```
Comment (UdjinM6) [inline test/functional/p2p_invalid_messages.py:9]: 19304: should be marked partial: missing changes in `test_large_inv` which we don't have yet

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/mempool_compatibility.py (hunk 9 above)
Comment (vijaydasmp) [reply test/functional/mempool_compatibility.py]: removed it

File: test/functional/test_framework/messages.py (hunk 10 above)
Comment (vijaydasmp) [reply test/functional/test_framework/messages.py:1838]: corrected

File: test/functional/p2p_invalid_messages.py (hunk 11 above)
Comment (vijaydasmp) [reply test/functional/p2p_invalid_messages.py:9]: done

**Review by UdjinM6 — APPROVED** [review]
utACK

**Comment by vijaydasmp** [comment]
Hello @UdjinM6 @PastaPastaPasta  please review

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Comment by knst** [comment]
For bitcoin 19350 seems can be updated couple more files:
```
test/functional/wallet_basic.py:
        self.stop_node(0)
        self.start_node(0, extra_args=["-walletrejectlongchains", "-limitancestorcount=" + str(2 * chainlimit)])
```
Change to `restart_node(0)`

Also:
```
feature_llmq_chainlocks.py:        self.stop_node(0)
feature_llmq_chainlocks.py-        self.start_node(0)
```

```
wallet_upgradetohd.py:        self.stop_node(0)
wallet_upgradetohd.py-        self.start_node(0, extra_args=['-keypool=10'])
```


Otherwise LGTM

**Review by knst — DISMISSED** [review]
utACK

File: test/functional/test_framework/test_framework.py [hunk 12]
```diff
@@ -9,8 +9,8 @@
 import copy
 from _decimal import Decimal
 from enum import Enum
-import logging
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/test_framework.py:12]: 18828 should be marked as partial

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
everything seems fine here: only one issue

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

File: test/functional/test_framework/test_framework.py (hunk 12 above)
Comment (vijaydasmp) [reply test/functional/test_framework/test_framework.py:12]: marked 18828 as (partial)

**Comment by vijaydasmp** [comment]
resolved all comments Requesting @UdjinM6 @PastaPastaPasta @knst to please review/approve

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Comment by vijaydasmp** [comment]
Hello @PastaPastaPasta , please have a look

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**